### PR TITLE
Add Post "Introducing KWOK for Istio Control Plane Testing"

### PR DESCRIPTION
**Pull request by wzshiming**
Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

- [ ] Ambient
- [x] Docs
- [ ] Installation
- [ ] Networking
- [x] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [x] Test and Release
- [ ] User Experience
- [x] Developer Infrastructure

### CODE REVIEW

File: content/en/blog/2023/kwok/index.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+For the latest information, see [KWOK with Istio].
+
+To create a test environment for the Istio control plane with KWOK, you can follow these steps:
+
+### Set up Cluster
+
+``` bash
+kwokctl create cluster --runtime kind
+```
```
Comment (windsonsea) [inline content/en/blog/2023/kwok/index.md:50]:
```suggestion
{{< text bash >}}
$ kwokctl create cluster --runtime kind
{{< /text >}}
```

File: content/en/blog/2023/kwok/index.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+### Create Node
+
+``` bash
+kubectl apply -f https://kwok.sigs.k8s.io/examples/node.yaml
+```
```
Comment (windsonsea) [inline content/en/blog/2023/kwok/index.md:56]:
```suggestion
{{< text bash >}}
$ kubectl apply -f https://kwok.sigs.k8s.io/examples/node.yaml
{{< /text >}}
```

**Review by windsonsea — COMMENTED** [review]
Would you mind to use the code style in istio docs like:

```
{{< text bash >}}
$ kubectl apply -f {{< github_file >}}/samples/bookinfo/networking/destination-rule-all.yaml
{{< /text >}}
```

**Comment by wzshiming** [comment]
Done

File: content/en/blog/2023/kwok/index.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+title: "Introducing KWOK for Istio Control Plane Testing"
+description: "Looking for a reliable way to test your Istio control plane at scale? Have you ever wondered how you could set up a cluster of thousands of nodes, simulate real nodes with a low resource footprint, and test at scale without spending a lot of money on infrastructure? Look no further! KWOK makes it easy."
+publishdate: 2023-07-20
+attribution: "Shiming Zhang (DaoCloud)"
+keywords: [Istio, kwok, testing]
+---
+
+Looking for a reliable way to test your Istio control plane at scale?
+Have you ever wondered how you could set up a cluster of thousands of nodes, simulate real nodes with a low resource footprint, and test at scale without spending a lot of money on infrastructure?
+Look no further! [KWOK] makes it easy.
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:11]:
```suggestion
Look no further! [KWOK](https://kwok.sigs.k8s.io/) makes it easy.
```

File: content/en/blog/2023/kwok/index.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+- **Flexibility**: You can configure different node types, labels, taints, capacities, conditions, etc., and you can configure different pod behaviors, status, etc. to test different scenarios and edge cases.
+
+{{< image width="90%"
+    link="./manage-clusters.svg"
+    caption="Using kwokctl to manage simulated clusters"
+    >}}
+
+## Testing Istio with KWOK
+
+For the latest information, see [KWOK with Istio].
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:42]:
```suggestion
For the latest information, see [KWOK with Istio](https://kwok.sigs.k8s.io/docs/examples/istio/).
```

File: content/en/blog/2023/kwok/index.md [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+### Deploy Istio
+
+{{< text bash >}}
+$ istioctl install -y
+{{< /text >}}
+
+### Migrate Controllers to Real Node
+
+We need to use [KIND] to run the controller because [KWOK] itself cannot actually run a Pod.
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:66]:
```suggestion
We need to use [KIND](https://kind.sigs.k8s.io/) to run the controller because [KWOK](https://kwok.sigs.k8s.io/) itself cannot actually run a Pod.
```

File: content/en/blog/2023/kwok/index.md [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+## Limitations
+
+Currently, we can only simulate the cluster data, but the real environment is actually connected with Sidecar's xDS, which cannot be simulated by KWOK.
+
+So we need a KWOK-based xDS client generator to complete the stress simulation, which is not yet available.
+
+## Getting Involved
+
+If you're interested in participating in future discussions or development related to [KWOK], there are several ways to get involved:
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:90]:
```suggestion
If you're interested in participating in future discussions or development related to [KWOK](https://kwok.sigs.k8s.io/), there are several ways to get involved:
```

File: content/en/blog/2023/kwok/index.md [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+
+- Slack: [#kwok] for general usage discussion, [#kwok-dev] for development discussion. (visit [slack.k8s.io] for a workspace invitation)
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:92]:
```suggestion
- Slack: [#kwok](https://kubernetes.slack.com/messages/kwok/) for general usage discussion, [#kwok-dev](https://kubernetes.slack.com/messages/kwok-dev/) for development discussion. (visit [slack.k8s.io](https://slack.k8s.io/) for a workspace invitation)
```

File: content/en/blog/2023/kwok/index.md [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+- Open Issues/PRs/Discussions in [sigs.k8s.io/kwok]
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:93]:
```suggestion
- Open Issues/PRs/Discussions in [sigs.k8s.io/kwok](https://sigs.k8s.io/kwok/)
```

File: content/en/blog/2023/kwok/index.md [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+We welcome feedback and contributions from anyone who would like to join us in this exciting project.
+
+[sigs.k8s.io/kwok]: https://sigs.k8s.io/kwok/
+[#kwok]: https://kubernetes.slack.com/messages/kwok/
+[#kwok-dev]: https://kubernetes.slack.com/messages/kwok-dev/
+[slack.k8s.io]: https://slack.k8s.io/
+[KWOK]: https://kwok.sigs.k8s.io/
+[KIND]: https://kind.sigs.k8s.io/
+[KWOK with Istio]: https://kwok.sigs.k8s.io/docs/examples/istio/
+[Istio in Kind]: https://istio.io/latest/docs/setup/platform-setup/kind/
```
Comment (wilsonwu) [inline content/en/blog/2023/kwok/index.md:104]:
```suggestion
```

**Review by wilsonwu — COMMENTED** [review]
Seems istio doc need put links in content directly.

**Comment by ericvn** [comment]
For the lint errors, which in this case are spelling errors, you can click on the `Details` after the failing test, an then click on the build.log link to see the log. Unfortunately, the logs aren't color-coded like they are when run normally so it may be hard to figure out which word is misspelled. Else, run `make lint` locally and the misspelled words are highlighted.

**Comment by ericvn** [comment]
You can see https://istio.io/v1.16/docs/releases/contribute/build/#test-your-changes for more info.

File: content/en/blog/2023/kwok/index.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+$ kubectl apply -f {{< github_file >}}/samples/bookinfo/platform/kube/bookinfo.yaml
+{{< /text >}}
+
+You will see that the Pod is running on a fake node. However, it is still an injected sidecar.
+
+## Limitations
+
+Currently, we can only simulate the cluster data, but the real environment is actually connected with Sidecar's xDS, which cannot be simulated by [KWOK].
+
+So we need a [KWOK] based xDS client connect generator to complete the stress simulation, which is not yet available.
```
Comment (wzshiming) [inline content/en/blog/2023/kwok/index.md:86]: I made a component [pwod](https://github.com/wzshiming/pwod) that generates xds connections based on pod data from a kwok cluster and it feels good!

**Comment by wzshiming** [comment]
> You can see https://istio.io/v1.16/docs/releases/contribute/build/#test-your-changes for more info.

Done

> Seems istio doc need put links in content directly.

Both should be acceptable, right?

**Comment by ericvn** [comment]
https://istio.io/latest/docs/releases/contribute/shortcodes/#add-links-to-other-pages

**Comment by ericvn** [comment]
And not sure why my prior link was the 1.16 link, probably because the search picked that one, and I just copied/pasted.

File: content/en/blog/2023/kwok/index.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+: `kwokctl` is a CLI tool designed to streamline the creation and management of clusters, with nodes simulated by `kwok`.
+
+## Why use KWOK?
+
+KWOK has several advantages:
+
+- **Lightweight**: You can simulate thousands of nodes on your laptop without significant consumption of CPU or memory resources.
+- **Fast**: You can create and delete clusters and nodes almost instantly, without waiting for boot or provisioning.
+- **Compatibility**: [KWOK] works with any tools or clients that are compliant with Kubernetes APIs, such as kubectl, helm, etc.
+- **Portability**: [KWOK] has no specific hardware or software requirements. You can run it using prebuilt images, once Docker/Podman/Nerdctl is installed. Alternatively, binaries are also available for all platforms and can be easily installed.
```
Comment (dhawton) [inline content/en/blog/2023/kwok/index.md:32]:
```suggestion
- **Portability**: [KWOK] has no specific hardware or software requirements. You can run it using prebuilt images, once Docker/Podman/nerdctl is installed. Alternatively, binaries are also available for all platforms and can be easily installed.
```

File: content/en/blog/2023/kwok/index.md [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+In this article, we'll introduce KWOK and show you how to test Istio with KWOK.
+
+## What is KWOK?
+
+KWOK stands for Kubernetes WithOut Kubelet. So far, it provides two tools:
+
+`kwok`
+: `kwok` is the cornerstone of this project, responsible for simulating the lifecycle of
+fake nodes, pods, and other Kubernetes API resources.
```
Comment (ericvn) [inline content/en/blog/2023/kwok/index.md:22]: Looking at the page, this section seems like it could be formatted differently. Just placing the tools above the description looks out of place. Maybe as a bullet point with a description like the advantages.

File: content/en/blog/2023/kwok/index.md [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+  and you can configure different pod behaviors, status, etc. to test different scenarios and edge cases.
+
+{{< image width="90%"
+    link="./manage-clusters.svg"
+    caption="Using kwokctl to manage simulated clusters"
+    >}}
+
+## Testing Istio with KWOK
+
+For the latest information, see [KWOK with Istio](https://kwok.sigs.k8s.io/docs/examples/istio/).
```
Comment (ericvn) [inline content/en/blog/2023/kwok/index.md:51]: The information here is copy of the link. Maybe move this paragraph inside a Tip section and place after the next paragraph.

File: content/en/blog/2023/kwok/index.md [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+{{< /text >}}
+
+You will see that the Pod is running on a fake node. However, it is still an injected sidecar.
+
+## Limitations
+
+With KWOK alone, we can only simulate the cluster data,
+but the real environment is actually also connected with Sidecar's xDS, which cannot be simulated.
+So we need a KWOK based xDS client connection generator to complete the load simulation,
+I made a [PWOD](https://github.com/wzshiming/pwod) to simulate the xDS client connection.
```
Comment (ericvn) [inline content/en/blog/2023/kwok/index.md:96]:
```suggestion
With KWOK alone, we can only simulate the cluster data,
while the real environment is also connected with Sidecar's xDS, which cannot be simulated.
We need a KWOK based xDS client connection generator to complete the load simulation, so
I made [PWOD](https://github.com/wzshiming/pwod) to simulate the xDS client connection.
```

**Review by ericvn — COMMENTED** [review]
A few comments.

File: content/en/blog/2023/kwok/index.md (hunk 14 above)
Comment (howardjohn) [inline content/en/blog/2023/kwok/index.md:96]: Saying you _made_ this feels a bit disingenuous. As far as I can tell, what was done is a library that *I made*, https://github.com/howardjohn/pilot-load/, was both directly copy+pasted and imported directly, without attribution. The whole library is 300 lines of code with virtually every line identically copied from `pilot-load`.

I am happy to see more work in performance testing area, and happy to collaborate in that regard. While I am sure you had good intentions this feels more like stealing my work and passing it off as your own (and then publicizing it on a blog that I am heavily involved in...).

Would it make sense to just directly use `pilot-load` here? As far as I can tell `pwod` is just a strict subset of `pilot-load` -- its not clear why another project is needed, even ignoring attribution issues.

Technically you could replace the entire blog with `pilot-load`, as it can do the work of `kwok` for the purpose of Istio testing... but I can understand why that would not be desired :slightly_smiling_face:

**Review by howardjohn — CHANGES_REQUESTED** [review]
.

File: content/en/blog/2023/kwok/index.md (hunk 14 above)
Comment (craigbox) [reply content/en/blog/2023/kwok/index.md:96]: image

Comment (wzshiming) [reply content/en/blog/2023/kwok/index.md:96]: > Saying you made this feels a bit disingenuous. As far as I can tell, what was done is a library that I made, https://github.com/howardjohn/pilot-load/, was both directly copy+pasted and imported directly, without attribution. The whole library is 300 lines of code with virtually every line identically copied from pilot-load.

The demo project is something I realized I missed while making this post.
The `go.mod` of the project in it references pilot-load directly and then rewrites the entry point, referencing only a few needed packages, for a quick implementation of my idea.

I [left a comment on your issue](https://github.com/howardjohn/pilot-load/issues/9#issuecomment-1641466182) almost the same day I made the demo to get your opinion.

> I am happy to see more work in performance testing area, and happy to collaborate in that regard. While I am sure you had good intentions this feels more like stealing my work and passing it off as your own (and then publicizing it on a blog that I am heavily involved in...).

I apologize for making you feel offended. 🙏🙏🙏

> Would it make sense to just directly use pilot-load here? As far as I can tell pwod is just a strict subset of pilot-load -- its not clear why another project is needed, even ignoring attribution issues.
> Technically you could replace the entire blog with pilot-load, as it can do the work of kwok for the purpose of Istio testing... but I can understand why that would not be desired 🙂

I agree with you.

3 years ago when I was actually doing a production press test for Istio, and I ended up abstracting two parts of the logic from the test, [wzshiming/xds](https://github.com/wzshiming/xds) and [fake-kubelet](https://github.com/wzshiming/fake-kubelet/tree/v0.2.0), originally I wanted to use the xds for the demo, but I felt that the dependency hadn't been maintained in 3 years, so I referenced the xds from pilot-load.

I took over this test task from someone else, that code that was hard-coded to inject test data into the Istio source code for the purpose of pressure testing, and I spent a lot of time learning about Istio and completely rewriting the test code.

Then I realized that most of the testers using it weren't interested in understanding the principles behind it and just wanted to test it in the most intuitive and simple way possible.

Comment (howardjohn) [reply content/en/blog/2023/kwok/index.md:96]: I didn't mean to come off as super offended, mostly just surprised. I think kwok is a really great idea but it seems like the xds part can just be a part of pilot-load. I am happy to collaborate there, and possibly in this blog if you are interested, accept PRs there, even move it to a neutral location if desired. Let me know what interests you

Comment (wzshiming) [reply content/en/blog/2023/kwok/index.md:96]: of course，thank you

Comment (wzshiming) [reply content/en/blog/2023/kwok/index.md]: Could you add something that describes how to use the pilot-load with kwok?

Comment (howardjohn) [reply content/en/blog/2023/kwok/index.md]: Just need to sue `pilot-load cluster`. The `--cluster-type` has 3 modes:

* real: actually create real pods. Can be combined with kwok to make those 0 cost
* fake-node: replace kwok basically. Create real pods but assign to fake nodes
* Fake: connect to a plain apiserver with no controller-manager at all

File: content/en/blog/2023/kwok/index.md [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+
+We can use `pilot-load cluster --cluster-type real` with `kwokctl create cluster` to simulate a complete Istio control plane performance test.
+
+## Look ahead
+
+I have an idea for `pilot-load` to support `kwok` directly, but it is not implemented yet.
+
+That is `pilot-load` watches all the pods managed by kwok and emulates the xDS client.
+
+Then we can then manipulate the resource, view the metrics panel and stress test it as we normally would with Istio.
```
Comment (wzshiming) [inline content/en/blog/2023/kwok/index.md:111]: @howardjohn

This is what I did in that demo project, it's very simple in principle and code, but it's a great experience.

**Comment by wzshiming** [comment]
I've updated it as per the comments.

**Comment by rootsongjc** [comment]
I found this article constructive and useful for testing Istio, and I'm also using KWOK, which is really handy compared to Kind and other Kubernetes emulation tools.

**Review by craigbox — COMMENTED** [review]
Hey @wzshiming - this seems like something we should get published.  Are you open to another quick round of review to get this live?  Happy to help.
